include/nvexec/detail/memory.cuh

```diff
@@ -110,4 +116,243 @@ namespace nvexec::STDEXEC_STREAM_DETAIL_NS {
       return make_host<__decay_t<A>>(status, resource, (A&&) t);
     }
 
+  struct pinned_resource : public std::pmr::memory_resource {
+    pinned_resource() noexcept {
+    }
+
+    void* do_allocate(size_t bytes, size_t /* alignment */) override {
+      void* ret;
+
+      if (cudaError_t status = STDEXEC_DBG_ERR(cudaMallocHost(&ret, bytes));
+          status != cudaSuccess) {
+        throw std::bad_alloc();
+      }
+
+      return ret;
+    }
+
+    void do_deallocate(void* ptr, size_t /* bytes */, size_t /* alignment */) override {
+      STDEXEC_DBG_ERR(cudaFreeHost(ptr));
+    }
+
+    bool do_is_equal(const std::pmr::memory_resource& other) const noexcept override {
+      return this == &other;
+    }
+  };
+
+  struct gpu_resource : public std::pmr::memory_resource {
+    gpu_resource() noexcept {
+    }
+
+    void* do_allocate(size_t bytes, size_t /* alignment */) override {
+      void* ret;
+
+      if (cudaError_t status = STDEXEC_DBG_ERR(cudaMalloc(&ret, bytes)); status != cudaSuccess) {
+        throw std::bad_alloc();
+      }
+
+      return ret;
+    }
+
+    void do_deallocate(void* ptr, size_t /* bytes */, size_t /* alignment */) override {
+      STDEXEC_DBG_ERR(cudaFree(ptr));
+    }
+
+    bool do_is_equal(const std::pmr::memory_resource& other) const noexcept override {
+      return this == &other;
+    }
+  };
+
+  struct managed_resource : public std::pmr::memory_resource {
+    void* do_allocate(size_t bytes, size_t /* alignment */) override {
+      void* ret;
+
+      if (cudaError_t status = STDEXEC_DBG_ERR(cudaMallocManaged(&ret, bytes));
+          status != cudaSuccess) {
+        throw std::bad_alloc();
+      }
+
+      return ret;
+    }
+
+    void do_deallocate(void* ptr, size_t /* bytes */, size_t /* alignment */) override {
+      STDEXEC_DBG_ERR(cudaFree(ptr));
+    }
+
+    bool do_is_equal(const std::pmr::memory_resource& other) const noexcept override {
+      return this == &other;
+    }
+  };
+
+  struct monotonic_buffer_resource : std::pmr::memory_resource {
+    static constexpr size_t block_alignment = 256;
+
+    struct block_descriptor_t {
+      void* ptr{};
+      size_t total{};
+    };
+
+    std::pmr::memory_resource* upstream;
+    std::vector<block_descriptor_t> allocated_blocks;
```

Comment:
Would there still be allocations for each node in the list? The two operations involving this list are pushing a new block and traversing it in the destructor. Pushing a new block into a linked list is O(N) allocations, while it's amortized O(1) for the vector. Traversing the linked list is accessing N random locations, whereas, for the vector, it will be predictable and local. Could you please elaborate on how the locality is improved by linked list usage?

Reply:
What I mean by "intrusive linked list" is for each block allocation to store a pointer to the next block allocation itself, rather than having a separate data structure storing pointers to all the blocks.

Reply:
Just to make sure I understand. You haven't suggested making an intrusive list of the `block_descriptor_t` but rather add the descriptor to the memory allocation itself. I'm afraid this defines the purpose of replacing standard memory resources in the first place. Memory resources provided as part of this PR should guarantee that the memory allocated by upstream resource is not accessed.